### PR TITLE
rpcwallet: 'ischange' field for 'getaddressinfo' RPC

### PR DESCRIPTION
**Pull request by rodentrabies**
Implementation of proposal in #14396.

This introduces `CWallet::IsChange(CScript&)` method and replaces original `CWallet::IsChange(CTxOut&)` method with overloaded version that delegates to the new method with *txout*'s `scriptPubKey`. In this way `TODO` note from the original method can still be addressed in a single place.

### CODE REVIEW

**Comment by promag** [comment]
Not sure about this.. What is the rationale?

Also, could test the new field.

**Review by promag — COMMENTED** [review]
@instagibbs because it can be misleading? Especially if you `fundrawtransaction` with explicit `changeAddress`

> especially when using things like importmulti where this can be set.

I'm not following, can you explain?

**Comment by achow101** [comment]
>> especially when using things like importmulti where this can be set.

> I'm not following, can you explain?

`importmulti` lets you mark imported things as internal, i.e. change. It can be useful to know whether something is change, especially if it was imported and so other ways of identifying change addresses may not be useful.

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -3540,6 +3540,7 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
             "  \"ismine\" : true|false,        (boolean) If the address is yours or not\n"
             "  \"iswatchonly\" : true|false,   (boolean) If the address is watchonly\n"
             "  \"isscript\" : true|false,      (boolean) If the key is a script\n"
+            "  \"ischange\" : true|false,      (boolean) If the address is for change output\n"
```
[the remark this answers is not in the document]
Comment (mrwhythat) [reply src/wallet/rpcwallet.cpp]: Thanks for the suggestion. Fixed that.

File: src/wallet/wallet.h [hunk 2]
```diff
@@ -961,6 +961,7 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
     isminetype IsMine(const CTxOut& txout) const;
     CAmount GetCredit(const CTxOut& txout, const isminefilter& filter) const;
     bool IsChange(const CTxOut& txout) const;
+    bool IsChange(const CScript& dest) const;
```
Comment (practicalswift) [inline src/wallet/wallet.h]: Please make the naming between those two consistent:

```
bool CWallet::IsChange(const CScript& script) const { … }
bool IsChange(const CScript& dest) const;
```

Make both `script` or both `dest` :-)

Comment (mrwhythat) [reply src/wallet/wallet.h]: Done.

**Comment by mrwhythat** [comment]
I'm not sure it needs a dedicated test, so I added testing of `ischange` field to `wallet_address_types.py` functional test.

**Review by promag — COMMENTED** [review]
Agree that there's no need for a dedicated test.

File: test/functional/wallet_address_types.py [hunk 3]
```diff
@@ -159,6 +159,13 @@ def test_change_output_type(self, node_sender, destinations, expected_type):
         change_addresses = [d for d in output_addresses if d not in destinations]
         assert_equal(len(change_addresses), 1)
 
+        # Check if 'getaddressinfo' correctly marks change address as change:
+        for d in output_addresses:
+            if d in destinations:
```
[the remark this answers is not in the document]
Comment (mrwhythat) [reply test/functional/wallet_address_types.py]: Done, with some changes though:)

File: test/functional/wallet_address_types.py [hunk 4]
```diff
@@ -159,6 +159,13 @@ def test_change_output_type(self, node_sender, destinations, expected_type):
         change_addresses = [d for d in output_addresses if d not in destinations]
         assert_equal(len(change_addresses), 1)
 
+        # Check if 'getaddressinfo' correctly marks change address as change:
```
[the remark this answers is not in the document]
Comment (mrwhythat) [reply test/functional/wallet_address_types.py]: Moved it to `wallet_basic.py` with some additional code to ensure transaction with change.

**Comment by sipa** [comment]
utACK 9298081df8a9b538cf5b572fd99f7063a67dae76

No reason why this can't be exposed.

**Review by promag — COMMENTED** [review]
Could test that setting a label for a change address sets `ischange` to false. This is the test change:

```diff
@@ -497,7 +497,12 @@ class WalletTest(BitcoinTestFramework):
         output_addresses = [vout['scriptPubKey']['addresses'][0] for vout in tx["vout"]]
         assert len(output_addresses) > 1
         for address in output_addresses:
-            assert_equal(self.nodes[0].getaddressinfo(address)['ischange'], address != destination)
+            ischange = self.nodes[0].getaddressinfo(address)['ischange']
+            assert_equal(ischange, address != destination)
+            if ischange: change = address
+        self.nodes[0].setlabel(change, 'foobar')
+        assert_equal(self.nodes[0].getaddressinfo(change)['ischange'], False)

 if __name__ == '__main__':
     WalletTest().main()
```
Also could add a small release note? Maybe in `doc/release-notes-14282.md`?

Tested ACK 9298081.

**Comment by mrwhythat** [comment]
@promag done and done. Let me know if release note is ok.

**Comment by promag** [comment]
utACK 14a0652.

**Comment by MeshCollider** [comment]
utACK https://github.com/bitcoin/bitcoin/commit/14a06525b2ed41d9a549e744019c06afeef9b0c5

**Comment by ismail120572** [comment]
Ismail

Pada tanggal Sen, 5 Nov 2018 01:30 João Barbosa <notifications@github.com
menulis:

**Pull request merged**